File: Baking App/app/src/main/java/com/madonasyombua/bakingapp/adapters/RecipeListAdapter.java
package com.madonasyombua.bakingapp.adapters;

import android.annotation.SuppressLint;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.madonasyombua.bakingapp.R;

import com.madonasyombua.bakingapp.models.Ingredients;
import com.madonasyombua.bakingapp.models.Recipe;
import com.madonasyombua.bakingapp.helpers.Listeners;

import java.util.Locale;

import butterknife.BindView;
import butterknife.ButterKnife;


public class RecipeListAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private final Recipe mRecipe;
    private final Listeners.OnItemClickListener mOnItemClickListener;

    //constructor takes Recipe and a Listener
    public RecipeListAdapter(Recipe recipe, Listeners.OnItemClickListener onItemClickListener) {
        this.mRecipe = recipe;
        this.mOnItemClickListener = onItemClickListener;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == 0) {
            return new RecipeListViewHolder(LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.recipe_ingredient_list_item, parent, false));
        } else {
            return new StepsAdapter.StepViewHolder(LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.recipe_step_list_item, parent, false));
        }

    }

    @SuppressLint({"RecyclerView", "SetTextI18n"})
    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, final int position) {
        if (holder instanceof RecipeListViewHolder) {
            RecipeListViewHolder viewHolder = (RecipeListViewHolder) holder;

            StringBuilder ingValue = new StringBuilder();
            for (int i = 0; i < mRecipe.getIngredients().size(); i++) {
                Ingredients ingredients = mRecipe.getIngredients().get(i);
                ingValue.append(String.format(Locale.getDefault(), "• %s (%d %s)", ingredients.getIngredient(), ingredients.getQuantity(), ingredients.getMeasure()));
                if (i != mRecipe.getIngredients().size() - 1)
                    ingValue.append("\n");
            }

            viewHolder.ingredientsTv.setText(ingValue.toString());
        } else if (holder instanceof StepsAdapter.StepViewHolder) {
            StepsAdapter.StepViewHolder viewHolder = (StepsAdapter.StepViewHolder) holder;
            viewHolder.mTvStepOrder.setText(String.valueOf(position - 1) + ".");
            viewHolder.mTvStepName.setText(mRecipe.getSteps().get(position - 1).getShortDescription());

            holder.itemView.setOnClickListener(v -> {
                if (mOnItemClickListener != null)
                    mOnItemClickListener.onItemClick(position - 1);
            });
        }
    }


    @Override
    public int getItemViewType(int position) {
        if (position == 0)
            return 0;
        else
            return 1;
    }

    @Override
    public int getItemCount() {
        return mRecipe.getSteps().size() + 1;
    }

    public class RecipeListViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.ingredients_text)
        public TextView ingredientsTv;

        RecipeListViewHolder(View itemView) {
            super(itemView);

            ButterKnife.bind(this, itemView);

        }

    }

}

File: Baking App/app/src/main/java/com/madonasyombua/bakingapp/adapters/RecipeListAdapter.kt
/*
 * Copyright (C) 2018 Madonah Syombua
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.madonasyombua.bakingapp.adapters

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import com.madonasyombua.bakingapp.R
import com.madonasyombua.bakingapp.models.Recipe
import java.util.*

class RecipeListAdapter     //constructor takes Recipe and a Listener
(private val mRecipe: Recipe, private val mOnItemClickListener: (Int) -> Unit) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if (viewType == 0) {
            RecipeListViewHolder(LayoutInflater.from(parent.context)
                    .inflate(R.layout.recipe_ingredient_list_item, parent, false))
        } else {
            StepsAdapter.StepViewHolder(LayoutInflater.from(parent.context)
                    .inflate(R.layout.recipe_step_list_item, parent, false))
        }
    }

    @SuppressLint("RecyclerView", "SetTextI18n")
    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is RecipeListViewHolder) {
            val ingValue = StringBuilder()
            for (i in mRecipe.ingredients.indices) {
                val ingredients = mRecipe.ingredients[i]
                ingValue.append(String.format(Locale.getDefault(), "• %s (%d %s)", ingredients?.ingredient, ingredients?.quantity, ingredients?.measure))
                if (i != mRecipe.ingredients.size - 1) ingValue.append("\n")
            }
            holder.ingredientsTv?.text = ingValue.toString()
        } else if (holder is StepsAdapter.StepViewHolder) {
            val viewHolder: StepsAdapter.StepViewHolder = holder
            viewHolder.mTvStepOrder.text = (position - 1).toString() + "."
            viewHolder.mTvStepName.text = (mRecipe.steps[position - 1]?.shortDescription ?: holder.itemView.setOnClickListener {
                v: View? -> mOnItemClickListener?.onItemClick(position - 1) }) as CharSequence?
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == 0) 0 else 1
    }

    override fun getItemCount(): Int {
        return mRecipe.steps.size + 1
    }

    inner class RecipeListViewHolder internal constructor(itemView: View?) : RecyclerView.ViewHolder(itemView!!) {
        @BindView(R.id.ingredients_text)
        var ingredientsTv: TextView? = null

        init {
            ButterKnife.bind(this, itemView!!)
        }
    }
}

private fun Any.onItemClick(i: Int) {

}
